diffrax/_step_size_controller/constant.py

```diff
@@ -47,15 +52,40 @@ def adapt_step_size(
         args: Args,
         y_error: Y | None,
         error_order: RealScalarLike | None,
-        controller_state: RealScalarLike,
-    ) -> tuple[bool, RealScalarLike, RealScalarLike, bool, RealScalarLike, RESULTS]:
+        controller_state: tuple[IntScalarLike, IntScalarLike, RealScalarLike, RealScalarLike]
+    ) -> tuple[bool, 
+            RealScalarLike,
+            RealScalarLike,
+            bool,
+            tuple[IntScalarLike, IntScalarLike, RealScalarLike, RealScalarLike], 
+            RESULTS
+        ]:
         del t0, y0, y1_candidate, args, y_error, error_order
+        step, num_steps, t0_sim, t1_sim = controller_state
+        step += 1
+        eqx.error_if(step,
+             step > num_steps,
+             "ConstantStepSize exceeded total number of expected steps, this may be because "
+             "the simulation continued after reaching t1 due to a floating point error. "
+             "If this occurs please file a bug report along with a MWE "
+             "at https://github.com/patrick-kidger/diffrax/issues so we can look into it."
+        )
+
+        time_dtype = jnp.result_type(t0_sim, t1_sim)
+
+        # Calculate step size by calculating fraction of `t1 - t0` to avoid compounding
+        # of truncation/rounding errors
+        t1_next = lax.cond(step == num_steps,
```

Comment:
`cond` actually has a bit more overhead than a `jnp.where`, and in this case it would only save work once in the whole solve. For that reason I think a `where` would be better here.